### PR TITLE
Add topic endpoints

### PR DESCRIPTION
**Pull request by cw9**

### CODE REVIEW

File: src/query/api/v1/handler/topic/add.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+// AddHandler is the handler for topic adds.
+type AddHandler Handler
+
+// NewAddHandler returns a new instance of AddHandler.
+func NewAddHandler(client clusterclient.Client, cfg config.Configuration) *AddHandler {
+	return &AddHandler{client: client, cfg: cfg, fn: Service}
+}
+
+func (h *AddHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/add.go]: var()

File: src/query/api/v1/handler/topic/add.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	logger := logging.WithContext(ctx)
+
+	var req admin.TopicAddRequest
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/add.go]: var()

File: src/query/api/v1/handler/topic/add.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	rErr := parseRequest(r, &req)
+	if rErr != nil {
+		logger.Error("unable to parse request", zap.Any("error", rErr))
+		xhttp.Error(w, rErr.Inner(), rErr.Code())
+		return
+	}
+
+	service, err := h.fn(h.client, r.Header)
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/add.go]: can you rename this serviceFn

File: src/query/api/v1/handler/topic/add.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		logger.Error("unable to get service", zap.Any("error", err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	t, err := service.Get(topicName(r))
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/add.go]: can `t` be nil? Do you need to handle that

File: src/query/api/v1/handler/topic/common.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+type serviceFn func(clusterClient clusterclient.Client, headers http.Header) (topic.Service, error)
+
+// Handler represents a generic handler for topic endpoints.
+type Handler struct {
+	// This is used by other topic Handlers
+	// nolint: structcheck
+	client clusterclient.Client
+	cfg    config.Configuration
+
+	fn serviceFn
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/common.go]: can you rename this serviceFn

File: src/query/api/v1/handler/topic/common.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// Service gets a topic service from m3cluster client
+func Service(clusterClient clusterclient.Client, headers http.Header) (topic.Service, error) {
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/common.go]: Do you need to accept headers?

File: src/query/api/v1/handler/topic/common.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+// RegisterRoutes registers the topic routes
+func RegisterRoutes(r *mux.Router, client clusterclient.Client, cfg config.Configuration) {
+	logged := logging.WithResponseTimeLogging
+
+	r.HandleFunc(InitURL, logged(NewInitHandler(client, cfg)).ServeHTTP).Methods(InitHTTPMethod)
+	r.HandleFunc(GetURL, logged(NewGetHandler(client, cfg)).ServeHTTP).Methods(GetHTTPMethod)
+	r.HandleFunc(AddURL, logged(NewAddHandler(client, cfg)).ServeHTTP).Methods(AddHTTPMethod)
+}
+
+func topicName(r *http.Request) string {
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/common.go]: Isnt the request a proto? How is this expected to be used. Can we just put it in the proto fields?

File: src/query/api/v1/handler/topic/get.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	logger := logging.WithContext(ctx)
+
+	service, err := h.fn(h.client, r.Header)
+	if err != nil {
+		logger.Error("unable to get service", zap.Any("error", err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	tn := topicName(r)
```
Comment (richardartoul) [inline src/query/api/v1/handler/topic/get.go]: can you just call this name

File: src/query/api/v1/handler/topic/add.go (hunk 4 above)
Comment (cw9) [reply src/query/api/v1/handler/topic/add.go]: t can't be nil when err==nil

**Pull request merged**